llvm/lib/Target/X86/X86InstrMisc.td

```diff
@@ -561,6 +561,16 @@ def MOV64rm : RI<0x8B, MRMSrcMem, (outs GR64:$dst), (ins i64mem:$src),
                  [(set GR64:$dst, (load addr:$src))]>;
 }
 
+def : Pat<(or (and GR64:$dst, -256), 
```

Comment:
What about GR32?

Reply:
The patterns probably belong in X86InstrCompiler.td where we keep most of these kinds of patterns.

Reply:
32bit move has implicit zero extension, so won't be applicable.

Moved the change to X86InstrCompiler.td

Reply:
I meant GPR32 for the destination type

def : Pat<(or (and GPR32:$dst, -256), 
              (i32 (zextloadi8 addr:$src))),

Reply:
Sorry I misunderstood. Fixed now.

Also updated the test. Note that sub16_32 case does not yet produce the optimized code for i386 because the pattern change (due to arg passing).